More languages

mattkatz opened the pull request:
Ensure that we output each language in a file per markdown document. 

If there are multiple language blocks in one document, each goes to the right file.

If the language is python, output __init__.py package indicators along the output tree to a file.

Allow multiple language descriptions for markdown blocks to go to the same file extension ("csharp", "c#" and "cs" all output ".cs" files)

Commit:
test that non markdown extension is ignored

Commit:
simplify file write and make_directories

simple file writing is simpler with pathlib.Path.write_text

For make_directories, python2 is end of life. python3 doesn't require
blank __init__.py files, so we can rely on pathlib.path.mkdir

Commit:
add per language test and data

Commit:
remove python indicator, switch to language

If mkcodes supports many language blocks, we can make the default
language python. safe mode should remove that default and not output
languages that aren't specified.

2 tests not passing

Commit:
introduce default language, lang mappings

rather than assume python, let's just recognize that's the default
language for unknown blocsk. People can override that as needed if they
want to put in some other assumptions.

Commit:
collect and write per-language code blocks

as we pass through a document, get per-language codeblocks, then write
those out to the correct extensionfile. By default, everything is still
python - though it can be overridden

Commit:
make other languages work

Now the default language is python
codeblocks in other languages can be broken out into their own language
files. There is a way to add more languages with mappings, but a
language fenced codeblock will default to use that language as the
output extension

Commit:
add test for unmapped extension

Commit:
default to adding __init__.py files to directories

Python unittest discover still relies on __init__.py files and doesn't
support namespace packages. By default we can create these but allow for
their suppression if using other test frameworks.

Commit:
clean up __init__.py addition logic

Also remove some commented code, shorten comments

Commit:
test __init__.py paths

For non python paths, don't declare them to be python packages.

Commit:
correct spelling mistakes

also conform new test to self.call

Commit:
separate 3rd party imports with a blank line

ryneeverett committed:
cleanup

mattkatz commented:
Not sure why this is showing conflicts as they seem really clearly one-sided.

ryneeverett commented on mkcodes.py:
```diff
@@ -13,48 +13,79 @@
-def github_codeblocks(filepath, safe):
-    codeblocks = []
+# much easier to write the other names that an extension is known by
+ext_map = {
```
Is there a specification or documentation of these language/extension mappings?

mattkatz replied:
Nope. this is just my freehand mapping of versions that I've seen in syntax highlighters to the language they output. 
Given the wide variety of syntax highlighters out there I'd expect this to evolve. 
Any ideas about where to make this mapping more formal is definitely welcome.

mattkatz replied:
for context: https://github.com/Python-Markdown/markdown/blob/master/markdown/extensions/fenced_code.py
python markdown just takes whatever you write as the language and passes it off to the pygments highliter. Which works great for java and py, but doesn't work well with csharp. Rather than cover everything, I just wanted to make it really clear where to add extensions and the mappings for other languages.

ryneeverett replied:
For lack of a better option, I'm tempted to incorporate pygments' mappings. https://github.com/pygments/pygments/blob/master/pygments/lexers/_mapping.py

mattkatz replied:
Urgh - the file extensions are kind of secondary there.
I took a stab at this by chopping it up in vim: https://gist.github.com/mattkatz/8830cbf4e91d83e4ef29a5100bf9f601
It's LONG

mattkatz replied:
Just seems like a YAGNI moment - if there is increased appetite for lots of languages, that's a good time to refactor out lots of languages into a file with extension mappings - but if not, it's hard to check for acuracy or test or port over changes from pygments.

ryneeverett replied:
Fair enough. I'd probably want to script the generation of that ext_map.py file so it can be easily updated.

ryneeverett commented on mkcodes.py:
```diff
@@ -13,48 +13,79 @@
+                # does it have a language?
+                lang_match = re.match(codeblock_open_re, line)
+                if lang_match:
+                    language = lang_match.group(2)
```
What would you think about `lang_match.group(2).lower()` here to lower the number of variations we need in `ext_map`?

mattkatz replied:
love it.

ryneeverett commented:
> Not sure why this is showing conflicts as they seem really clearly one-sided.

I rebased on master.

ryneeverett commented on mkcodes.py:
```diff
@@ -13,48 +13,79 @@
+                    if not safe:
+                        # we can sub a default language if not safe
+                        language = language or default_lang
+                else:
```
It seems that this entire `else` block can be deleted as no-op. if there's no language specified then `language = None` is fine -- `language = language or default_lang` will pick up the `default_lang` if the safety is off.

ryneeverett committed:
Ignore capitalization of language strings.

ryneeverett committed:
Remove no-op else clause.

ryneeverett committed:
Document extension/language mappings.

ryneeverett commented:
I pushed some changes onto the rebase I did the other day. If you're going to push more to this branch you should pull first and then you wont re-push the conflicts.

ryneeverett committed:
Language should be reset even if --unsafe.

The safety flag refers to whether a default language is used. This would
have made it also accept the previous language used, which doesn't seem
right.

ryneeverett committed:
Add warning when unhinted code blocks are skipped.

ryneeverett commented on mkcodes.py:
```diff
@@ -79,43 +113,55 @@ def get_files(inputs):
-def main(inputs, output, github, safe, package_python):
+              help='Add __init__.py files to python dirs for test discovery')
+@click.option('--default_lang', default='py',
+              help='Assumed language for code blocks without language hits.')
```
There is a bit of dissonance between the `--safe` flag and the `--default-lang` option that we may want to resolve. Some ideas:

- Perhaps the use of both of these flags should raise an error? (I'd imagine there's an option to indicate incompatible arguments in the `click` api.)
- Perhaps we should remove the `--safe` flag in  favor of `--default_lang=` (no value). Maybe `--safe` should just be an alias for `--default_lang=`.

mattkatz replied:
sure - I lean towards --safe == default_lang==None
 
the motivation for safe seems to be "ignore code blocks that don't have a language!
default_lang's motivation is "assume the language is this if you don't know".
If there is no language and you don't know what extension to use, I can't imagine a scenario where you want to output the codes.  
If you have some directories where the language on unmarked codes is py and some directories where the default lang  is elm, you can indicate that easily with multiple runs, one for each directory.

If you have a single file with multiple code blocks where you don't know the language, you really just need to get the document fixed to indicate the languages...

ryneeverett commented on mkcodes.py:
```diff
@@ -79,43 +113,55 @@ def get_files(inputs):
+        return
+    if to_path.relative_to(from_path):
+        if from_path != to_path.parent:
+            add_inits_along_path(from_path, to_path.parent)
```
Could you explain why you changed this `add_inits_along_path` implementation?

mattkatz replied:
originally we added an __init__.py file to every dir from the output base directory to every leaf directory.
Now, we may have some directories that will not have any python files in them - we should not add an __init__.py to those directories as they are not python packages. 

Instead, when outputting a .py file we can reasonably just output an __init__.py from the base output directory to the directory containing a python file.

Since we are just using touch to create blank files, there's no real point in checking to see if they are already there.

mattkatz commented:
> I pushed some changes onto the rebase I did the other day. If you're going to push more to this branch you should pull first and then you wont re-push the conflicts.

Ok - I'm working off of a fork, so I'll figure that out. I think that just means doing a pull rebase from your master into my fork branch.

ryneeverett commented on mkcodes.py:
```diff
@@ -79,43 +113,55 @@ def get_files(inputs):
+    (to_path / '__init__.py').touch()
+    if to_path == from_path:
+        return
+    if to_path.relative_to(from_path):
```
What is the purpose of this `if` clause? Shouldn't `to_path.relative_to(from_path)` always be truthy? Is it just a sanity check?

mattkatz replied:
sanity check - if later we make a mistake and somehow invoke this with a non-relative path you don't want to go up and up until we get perm denied or start dropping __init__.py in the user directory etc.

ryneeverett committed:
Use dash for cli parameters and fix typo.

ryneeverett commented on mkcodes.py:
```diff
@@ -79,43 +113,55 @@ def get_files(inputs):
+    if to_path == from_path:
+        return
+    if to_path.relative_to(from_path):
+        if from_path != to_path.parent:
```
I would have thought this line would prevent putting `__init__.py` in the `outputbasedir`, but that doesn't seem to be the case. What is it's purpose?

mattkatz replied:
That is the purpose of it, but I should have added a test to verify it!
I'm having trouble getting the code that is in this pull request to rebase on, but I did this:

tests/test.py:
```py
def test_other_languages(self):
    # previous assertions
    self.assertFalse(self._output_path_exists('__init__.py'))
```

mkcodes.py:
```py
def add_inits_along_path(from_path, to_path):
    to_path = to_path.expanduser().resolve()
    from_path = from_path.expanduser().resolve()
    if to_path == from_path:
        return
    (to_path / '__init__.py').touch()
    if to_path.relative_to(from_path):
        add_inits_along_path(from_path, to_path.parent)
```

the new test fails on the old version of add_inits_along_path (verifying your point) and succeeds with the new version of the function.

ryneeverett commented on mkcodes.py:
```diff
@@ -79,43 +113,55 @@ def get_files(inputs):
+        return
+    if to_path.relative_to(from_path):
+        if from_path != to_path.parent:
+            add_inits_along_path(from_path, to_path.parent)
```
It seems like this function could be simplified to:

```py
def add_inits_along_path(from_path, to_path):
    to_path = to_path.expanduser().resolve()
    from_path = from_path.expanduser().resolve()
    (to_path / '__init__.py').touch()
    if to_path != from_path:
        add_inits_along_path(from_path, to_path.parent)
```

Any objections?

mattkatz replied:
I think relative_to is a cheap and fast safety check - the implementation is based on the string representation of the path. As the code is right now, it's very safe, but I do like some safety checks when recursing UP that I'm less concerned about in recursing down.

See my above code for slightly simpler, but avoids __init__.py in the output base directory and has a safety check.

mattkatz replied:
this is not a *strong* objection - if you like that version, it's fine and will work great.

mattkatz commented on tests/test.py:
```diff
@@ -140,6 +141,22 @@ def test_prefixed_deep_blocks(self):
+        self.assertTrue(self._output_path_exists('no_py_tree/test_clean.js'))
+        self.assertFalse(self._output_path_exists('no_py_tree/__init__.py'))
+        self.assertTrue(self._output_path_exists('pytree/test_buried.py'))
+        self.assertTrue(self._output_path_exists('pytree/__init__.py'))
```
```suggestion
        self.assertTrue(self._output_path_exists('pytree/__init__.py'))
        self.assertFalse(self._output_path_exists('__init__.py'))
```
to test that we don't output __init__.py in the base output directory

ryneeverett committed:
Pull sanity check out of if-statement.

ryneeverett committed:
Don't create __init__.py in base output directory.

ryneeverett committed:
Avoid abusing return as a mid-function "break".

I would consider this anti-pattern when avoidable and I find it harder to
understand.

ryneeverett committed:
Remove redundant test assertion.

ryneeverett commented:
This is a huge improvement, thanks for sticking with it!

mattkatz commented:
Just checked into see the merge! Awesome! Feels great to contribute.
